controller/api/destination/endpoint_translator.go

```diff
@@ -0,0 +1,171 @@
+	"github.com/linkerd/linkerd2-proxy-api/go/net"
+	"github.com/linkerd/linkerd2/controller/api/destination/watcher"
+	"github.com/linkerd/linkerd2/pkg/addr"
+	"github.com/linkerd/linkerd2/pkg/k8s"
+	logging "github.com/sirupsen/logrus"
+)
+
+const defaultWeight uint32 = 10000
+
+// endpointTranslator statisfies EndpointUpdateListener and translates updates
+// into Destination.Get messages.
+type endpointTranslator struct {
+	controllerNS        string
+	identityTrustDomain string
+	enableH2Upgrade     bool
+	labels              map[string]string
+	stream              pb.Destination_GetServer
+	stopCh              chan struct{}
+	log                 *logging.Entry
+}
+
+func newEndpointTranslator(
+	controllerNS string,
+	identityTrustDomain string,
+	enableH2Upgrade bool,
+	service watcher.ServiceID,
+	stream pb.Destination_GetServer,
+	log *logging.Entry,
+) *endpointTranslator {
+	log = log.WithFields(logging.Fields{
+		"component": "endpoint-converter",
+		"service":   fmt.Sprintf("%s/%s", service.Namespace, service.Name),
+	})
+	labels := map[string]string{
+		"namespace": service.Namespace,
+		"service":   service.Name,
+	}
+	stopCh := make(chan struct{})
+	return &endpointTranslator{controllerNS, identityTrustDomain, enableH2Upgrade, labels, stream, stopCh, log}
+}
+
+func (et *endpointTranslator) Add(set watcher.PodSet) {
+	addrs := []*pb.WeightedAddr{}
+	for _, address := range set {
+		wa, err := et.toWeightedAddr(address)
+		if err != nil {
+			et.log.Errorf("Failed to translate endpoints to weighted addr: %s", err)
+			continue
+		}
+		addrs = append(addrs, wa)
+	}
+
+	add := &pb.Update{Update: &pb.Update_Add{
+		Add: &pb.WeightedAddrSet{
+			Addrs:        addrs,
+			MetricLabels: et.labels,
+		},
+	}}
+
+	et.log.Debugf("Sending destination add: %+v", add)
+	if err := et.stream.Send(add); err != nil {
+		et.log.Errorf("Failed to send address update: %s", err)
+	}
+}
+
+func (et *endpointTranslator) Remove(set watcher.PodSet) {
+	addrs := []*net.TcpAddress{}
+	for _, address := range set {
+		tcpAddr, err := et.toAddr(address)
+		if err != nil {
+			et.log.Errorf("Failed to translate endpoints to addr: %s", err)
+			continue
+		}
+		addrs = append(addrs, tcpAddr)
+	}
+
+	remove := &pb.Update{Update: &pb.Update_Remove{
+		Remove: &pb.AddrSet{
+			Addrs: addrs,
+		},
+	}}
+
+	et.log.Debugf("Sending destination remove: %+v", remove)
+	if err := et.stream.Send(remove); err != nil {
+		et.log.Errorf("Failed to send address update: %s", err)
+	}
+}
+
+func (et *endpointTranslator) NoEndpoints(exists bool) {
+	et.log.Debugf("NoEndpoints(%+v)", exists)
+
+	u := &pb.Update{
+		Update: &pb.Update_NoEndpoints{
+			NoEndpoints: &pb.NoEndpoints{
+				Exists: exists,
+			},
+		},
+	}
+
+	et.log.Debugf("Sending destination no endpoints: %+v", u)
+	if err := et.stream.Send(u); err != nil {
+		et.log.Errorf("Failed to send address update: %s", err)
+	}
+}
+
+func (et *endpointTranslator) toAddr(address watcher.Address) (*net.TcpAddress, error) {
```

Comment:
this method doesn't need to hang off `endpointTranslator`. is it for consistency with `toWeightedAddr`?

Reply:
that's right.  these were originally one method which I split into two and it seemed reasonable to keep them together by hanging both off of endpointTranslator